motr/io_pargrp.c

```diff
@@ -464,73 +455,170 @@ static int pargrp_iomap_populate(struct pargrp_iomap      *map,
+ * (new data units and udpated parity units), so we compare only the
+ * number of units (pages) to be read.
+ *
+ * By default, the segments in index vector pargrp_iomap::pi_ivec
+ * are suitable for read-old approach. Hence the index vector is
+ * changed only if read-rest approach is selected.
+ */
+static int pargrp_iomap_select_ro_rr(struct pargrp_iomap *map,
+				     m0_bcount_t data_pages_nr,
+				     m0_bcount_t parity_pages_nr)
+{
+	int rc;
+	/*
+	 * In read-old the number of pages to be read is the same as
+	 * the number of pages to be written.
+	 *
+	 * TODO: Can use number of data_buf structures instead of using
+	 * indexvec_page_nr().
+	 */
+	uint64_t ro_pages_nr = iomap_page_nr(map) + parity_pages_nr;
+	/*
+	 * In read-rest the number of pages to be read is all data
+	 * pages which are not fully spanned by the io vector.
+	 */
+	uint64_t rr_pages_nr = data_pages_nr -
+	                       map->pi_ops->pi_fullpages_find(map);
+
+	if (rr_pages_nr < ro_pages_nr || map->pi_trunc_partial) {
+		M0_LOG(M0_DEBUG, "[%p]: RR selected", map->pi_ioo);
+		map->pi_rtype = PIR_READREST;
+		rc = map->pi_ops->pi_readrest(map);
+		if (rc != 0)
+			return M0_ERR(rc);
+	} else {
+		M0_LOG(M0_DEBUG, "[%p]: RO selected", map->pi_ioo);
+		map->pi_rtype = PIR_READOLD;
+		rc = map->pi_ops->pi_readold_auxbuf_alloc(map);
+	}
+
+	return M0_RC(rc);
+}
+
+/**
+ * Builds the iomap structure of row/columns for data/parity, one segment at
+ * a time.
+ * This is heavily based on m0t1fs/linux_kernel/file.c::pargrp_iomap_populate
+ *
+ * @param map[out] The map top populate.
+ * @param ivec The extents of the global file we will operate on.
+ * @param cursor Where in which extent we should start.
+ * @return 0 for sucess, -errno otherwise.
+ */
+static int pargrp_iomap_populate(struct pargrp_iomap      *map,
+				 struct m0_ivec_cursor    *cursor,
+				 struct m0_bufvec_cursor  *buf_cursor)
+{
+	int                       rc;
+	bool                      rmw = false;
+	m0_bcount_t               grpsize;
+	m0_bindex_t               grpstart;
+	m0_bindex_t               grpend;
+	struct m0_pdclust_layout *play;
+	struct m0_op_io          *ioo;
+	struct m0_op             *op;
+	struct m0_obj            *obj;
+	struct m0_client         *instance;
+
+	M0_PRE(map != NULL);
+
+	ioo =  map->pi_ioo;
+	obj =  ioo->ioo_obj;
+	op  = &ioo->ioo_oo.oo_oc.oc_op;
+	instance = m0__op_instance(op);
+
+	play     = pdlayout_get(ioo);
+	grpsize  = data_size(play);
+	grpstart = grpsize * map->pi_grpid;
+	grpend   = grpstart + grpsize;
+
+	M0_ENTRY("[%p] map=%p ivec=%p", ioo, map, cursor->ic_cur.vc_vec);
+
+	/*
+	 * For a write, if this map does not span the whole parity group,
+	 * it is a read-modify-write.
+	 *
+	 * Note (Sining): as Client objects don't have attribute SIZE,
+	 * rmw is true even when grpstart is larger than current object end.
+	 * Does this cause any problem if returned 'read' pages are not
+	 * zeroed?
```

Comment:
@siningwuseagate, is this note still actual?

Reply:
@andriytk I don't know for now if the service guarantees that those `non-existing` blocks will always return 0, if it is the case, you can remove the note.

Reply:
Looks like the service does return zero-ed pages for non-existing units. Here is the quick test from the SAGE cluster on 2+1 pool:

```
$ mcp -prof 0x7000000000000001:0x51a -hax 172.18.1.36@o2ib:12345:1:1 -ep 172.18.1.36@o2ib:12345:4:1 -proc 0x7200000000000001:0x436 -pool 0x6f00000000000001:0x466 -v /tmp/8k 0x123008:0x001
2021/04/13 14:52:25 mio.go:549: W: off=0 len=8192 bs=8192 gs=8192 speed=615 (Kbytes/sec)
$ mcp -prof 0x7000000000000001:0x51a -hax 172.18.1.36@o2ib:12345:1:1 -ep 172.18.1.36@o2ib:12345:4:1 -proc 0x7200000000000001:0x436 -pool 0x6f00000000000001:0x466 -osz 12 -v 0x123008:0x001 /tmp/8k+4k
2021/04/13 14:54:20 mio.go:621: R: off=0 len=12288 bs=16384 gs=8192 speed=2000 (Kbytes/sec)
14:00
$ dd if=/tmp/8k+4k of=/tmp/8k-cmp bs=8K count=1
1+0 records in
1+0 records out
8192 bytes (8.2 kB) copied, 0.000308617 s, 26.5 MB/s
$ cmp /tmp/8k /tmp/8k-cmp
$ dd if=/tmp/8k+4k of=/tmp/4k bs=4K skip=2 count=1
1+0 records in
1+0 records out
4096 bytes (4.1 kB) copied, 0.000289731 s, 14.1 MB/s
$ ls -l /tmp/4k
-rw-r--r-- 1 tkachuk1 jusers 4096 Apr 13 14:59 /tmp/4k
$ hexdump -C /tmp/4k
00000000  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00001000
```